train.py

```diff
@@ -188,38 +190,50 @@ def train(params, args, local_rank, world_rank, world_size):
-            args.tboard_writer.add_scalar('Avg iters per sec', step_count/(end-start), iters)
-            fig = generate_images([inp, tar, gen])
+            args.tboard_writer.add_scalar('Avg iters per sec', step_count/float(end-start), iters)
+            fig = generate_images([inp.to(torch.float32),
+                                   tar.to(torch.float32),
+                                   gen.to(torch.float32)])
             args.tboard_writer.add_figure('Visualization, t2m', fig, iters, close=True)
 
         val_start = time.time()
-        val_loss = []
+        val_loss = torch.zeros(1, device=device)
         val_rmse = torch.zeros((params.n_out_channels), dtype=torch.float32, device=device)
-        valid_steps = 0
+        val_samples = torch.zeros(1, device=device)
         model.eval()
 
-        with torch.no_grad():
-            for i, data in enumerate(val_data_loader, 0):
-                with autocast(enabled=params.amp_enabled, dtype=params.amp_dtype):
+        with torch.inference_mode():
+            with torch.no_grad():
+                for i, data in enumerate(tqdm(val_data_loader, desc="Validation progress", disable=(world_rank!=0))):
+
+                    # copy data to gpu
                     inp, tar = map(lambda x: x.to(device), data)
-                    gen = model(inp)
-                    loss = loss_func(gen, tar)
+
+                    # forward pass
+                    with autocast(enabled=params.amp_enabled, dtype=params.amp_dtype):
+                        gen = model(inp)
+                        loss = loss_func(gen, tar)
+
+                    # update metrics
+                    val_loss += loss
                     val_rmse += weighted_rmse(gen, tar)
-                    if params.distributed:
-                        torch.distributed.all_reduce(loss)
-                        torch.distributed.all_reduce(val_rmse)
-                        val_rmse /= world_size
-                    val_loss.append(loss.item()/world_size)
-                valid_steps += 1
-
-        val_rmse /= valid_steps # Avg validation rmse
-        val_end = time.time()
-        if world_rank==0:
-            logging.info('  Avg val loss=%f'%np.mean(val_loss))
-            logging.info('  Total validation time: {} sec'.format(val_end - val_start)) 
-            args.tboard_writer.add_scalar('Loss/valid', np.mean(val_loss), iters)
-            args.tboard_writer.add_scalar('RMSE(u10m)/valid', val_rmse.cpu().numpy()[0], iters)
-            args.tboard_writer.flush()
+                    val_samples += float(inp.shape[0])
+                
+                if params.distributed:
+                    torch.distributed.all_reduce(val_loss)
+                    torch.distributed.all_reduce(val_rmse)
+                    torch.distributed.all_reduce(val_samples)
+                    
+                val_rmse_avg = (val_rmse / val_samples)
```

Comment:
What is this doing exactly? val_rmse is already a batch averaged metric.

Reply:
yes, I changed that

Reply:
where exactly? it's averaged in metrics.py